[SYCLomatic] Update QuickSilver test case. (#630)
oneapi-src/SYCLomatic-test: third-party-programs/Velocity-Bench/QuickSilver/README.md

# Migration example: Migrate QuickSilver to SYCL version

> [SYCLomatic](https://github.com/oneapi-src/SYCLomatic) is a project to assist developers in migrating their existing code written in different programming languages to the SYCL* C++ heterogeneous programming model. It is an open source version of Intel® DPC++ Compatibility Tool.

Reviewer: ```suggestion
[SYCLomatic](https://github.com/oneapi-src/SYCLomatic) is a project to assist developers in migrating their existing code written in different programming languages to the SYCL* C++ heterogeneous programming model. It is an open-source version of the Intel® DPC++ Compatibility Tool.
```

> This file lists the detail steps to migrate CUDA version of [QuickSilver](https://github.com/oneapi-src/Velocity-Bench/tree/main/QuickSilver) to SYCL version with SYCLomatic. As follow table summaries the migration environment, software required and so on.

Reviewer: ```suggestion
This file lists the detailed steps to migrate CUDA version of [QuickSilver](https://github.com/oneapi-src/Velocity-Bench/tree/main/QuickSilver) to SYCL version with SYCLomatic. As follow table summarizes the migration environment, the software required, and so on.
```

 ```
   $ export SYCLomatic_HOME=/path/to/install/SYCLomatic
   $ mkdir $SYCLomatic_HOME
   $ cd $SYCLomatic_HOME
   $ wget https://github.com/oneapi-src/SYCLomatic/releases/download/20240203/linux_release.tgz   #Change the timestamp 20240203 to latest one
   $ tar xzvf linux_release.tgz
   $ source setvars.sh
   $ dpct --version #SYCLomatic version
 ```
 
 
For more information on configuring environment variables, see [Use the setvars Script with Linux*](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-linux-or-macos.html).
 
### 2 Generate the compilation database 
``` sh
$ cd ${QuickSilver_HOME}/CUDA/src
$ make clean
$ intercept-build make
$ ls compile_commands.json  # make sure compile_commands.json is generated
compile_commands.json
```
### 3 Migrate the source code and build script
```sh
# From the CUDA directory as root directory:
$ cd ${QuickSilver_HOME}/CUDA
$ dpct --in-root=. -p=./src/compile_commands.json --out-root=out --gen-build-script --cuda-include-path=/usr/local/cuda/include
```
Description of the options: 

>  * `--in-root`: provide input files to specify where to locate the CUDA files that needs migration.

Reviewer: ```suggestion
 * `--in-root`: provide input files to specify where to locate the CUDA files that need migration.
```

>  * `-p`: specify compilation database to migrate the whole project.

Reviewer: ```suggestion
 * `-p`: specify the compilation database to migrate the whole project.
```

MPI Version             : 3.0
Number of MPI ranks     : 1
Number of OpenMP Threads: 1
Number of OpenMP CPUs   : 1

Simulation:
   dt: 1e-08
   fMax: 0.1
   inputFile: ../../../Examples/AllScattering/scatteringOnly.inp
   energySpectrum: 
   boundaryCondition: octant
   loadBalance: 1
   cycleTimers: 0
   debugThreads: 0
   lx: 100
   ly: 100
   lz: 100
   nParticles: 10000000
   batchSize: 0
   nBatches: 10
   nSteps: 10
   nx: 10
   ny: 10
   nz: 10
   seed: 1029384756
   xDom: 0
   yDom: 0
   zDom: 0
   eMax: 20
   eMin: 1e-09
   nGroups: 230
   lowWeightCutoff: 0.001
   bTally: 1
   fTally: 1
   cTally: 1
   coralBenchmark: 0
   crossSectionsOut:

Geometry:
   material: sourceMaterial
   shape: brick
   xMax: 100
   xMin: 0
   yMax: 100
   yMin: 0
   zMax: 100
   zMin: 0

Material:
   name: sourceMaterial
   mass: 1000
   nIsotopes: 10
   nReactions: 9
   sourceRate: 1e+10
   totalCrossSection: 0.1
   absorptionCrossSection: flat
   fissionCrossSection: flat
   scatteringCrossSection: flat
   absorptionCrossSectionRatio: 0
   fissionCrossSectionRatio: 0
   scatteringCrossSectionRatio: 1

CrossSection:
   name: flat
   A: 0
   B: 0
   C: 0
   D: 0
   E: 1
   nuBar: 2.4

> WARNING: The enviroment variable SYCL_DEVICE_FILTER is deprecated. Please use ONEAPI_DEVICE_SELECTOR instead.

Reviewer: ```suggestion
WARNING: The environment variable SYCL_DEVICE_FILTER is deprecated. Please use ONEAPI_DEVICE_SELECTOR instead.
```

Figure Of Merit                  12.67 [Num Mega Segments / Cycle Tracking Time]
```
**Note:** 
* The testing result was collected when run on Intel(R) Core(TM) i7-13700K CPU backend with Intel® oneAPI Base Toolkit(2023.2 version).
* The Reference migrated code is attached in **migrated** folder.

If an error occurs during runtime, refer to [Diagnostics Utility for Intel® oneAPI Toolkits](https://www.intel.com/content/www/us/en/develop/documentation/diagnostic-utility-user-guide/top.html).

## QuickSilver License
[License.txt](https://github.com/oneapi-src/Velocity-Bench/blob/main/QuickSilver/LICENSE.md)

## Reference 
* Command Line Options of [SYCLomatic](https://oneapi-src.github.io/SYCLomatic/dev_guide/command-line-options-reference.html) or [Intel® DPC++ Compatibility Tool](https://software.intel.com/content/www/us/en/develop/documentation/intel-dpcpp-compatibility-tool-user-guide/top/command-line-options-reference.html)
* [oneAPI GPU Optimization Guide](https://www.intel.com/content/www/us/en/docs/oneapi/optimization-guide-gpu/)
* [SYCLomatic project](https://github.com/oneapi-src/SYCLomatic/)

## Trademarks information

> Intel, the Intel logo, and other Intel marks are trademarks of Intel Corporation or its subsidiaries.<br>

Reviewer: ```suggestion
Intel, the Intel logo, and other Intel marks are trademarks of Intel Corporation or its subsidiaries.
```

# Migration example: Migrate QuickSilver to SYCL version
[SYCLomatic](https://github.com/oneapi-src/SYCLomatic) is a project to assist developers in migrating their existing code written in different programming languages to the SYCL* C++ heterogeneous programming model. It is an open source version of Intel® DPC++ Compatibility Tool.

This file lists the detail steps to migrate CUDA version of [QuickSilver](https://github.com/oneapi-src/Velocity-Bench/tree/main/QuickSilver) to SYCL version with SYCLomatic. As follow table summaries the migration environment, software required and so on.

   | Optimized for         | Description
   |:---                   |:---
   | OS                    | Linux* Ubuntu* 22.04
   | Software              | Intel® oneAPI Base Toolkit, SYCLomatic
   | What you will learn   | Migration of CUDA code, Run SYCL code on oneAPI and Intel device
   | Time to complete      | 15 minutes

## Migrating QuickSilver to SYCL

### 1 Prepare the migration
#### 1.1 Get the source code of QuickSilver
```sh
   $ git clone https://github.com/oneapi-src/Velocity-Bench.git
   $ export QuickSilver_HOME=/path/to/Velocity-Bench/QuickSilver

>    $ cd ${QuickSilver_HOME}/CUDA/src

Reviewer: ```suggestion
 
```

Author: Done

   ```
      out/
      ├── MainSourceFiles.yaml
      ├── cudaImage.dp.cpp
      ├── cudaImage.h
      ├── QuickSilver.h
      ├── QuickSilver.h.yaml
      ├── QuickSilverD.dp.cpp
      ├── QuickSilverD.h
      ├── QuickSilverH.dp.cpp
      ├── QuickSilverH.h
      ├── cudautils.h
      ├── cudautils.h.yaml
      ├── geomFuncs.cpp
      ├── mainSift.cpp.dp.cpp
      └── matching.dp.cpp

   ```
### 4 Review the migrated source code and fix all `DPCT` warnings

SYCLomatic and [Intel® DPC++ Compatibility Tool](https://www.intel.com/content/www/us/en/developer/tools/oneapi/dpc-compatibility-tool.html) define a list of `DPCT` warnings and embed the warning in migrated source code if need manual effort to check. All the warnings in the migrated code should be reviewed and fixed. For detail of `DPCT` warnings and corresponding fix examples, refer to [Intel® DPC++ Compatibility Tool Developer Guide and Reference](https://www.intel.com/content/www/us/en/develop/documentation/intel-dpcpp-compatibility-tool-user-guide/top/diagnostics-reference.html) or [SYCLomatic doc page](https://oneapi-src.github.io/SYCLomatic/dev_guide/diagnostics-reference.html). 

> The icpx supports c++17, so you need to update:

Reviewer: ```suggestion
c++17 is required to build SYCL code, so you need to update -std option as follow:
```

```
$ grep "c++11" out/Makefile.dpct 
TARGET_0_FLAG_0 = -DHAVE_CUDA -DHAVE_UVM=1 -std=c++11 -O3 ${FLAGS}
TARGET_0_FLAG_1 = -DHAVE_CUDA -DHAVE_UVM=1 -std=c++11 -O3 ${FLAGS}
TARGET_0_FLAG_2 = -DHAVE_CUDA -DHAVE_UVM=1 -std=c++11 -O3 ${FLAGS}
TARGET_0_FLAG_7 = -DHAVE_CUDA -DHAVE_UVM=1 -std=c++11 -O3 ${FLAGS}
TARGET_0_FLAG_8 = -DHAVE_CUDA -DHAVE_UVM=1 -std=c++11 -O3 ${FLAGS}
TARGET_0_FLAG_11 = -DHAVE_CUDA -DHAVE_UVM=1 -std=c++11 -O3 ${FLAGS}
TARGET_0_FLAG_12 = -DHAVE_CUDA -DHAVE_UVM=1 -std=c++11 -O3 ${FLAGS}
TARGET_0_FLAG_14 = -DHAVE_CUDA -DHAVE_UVM=1 -std=c++11 -O3 ${FLAGS}
.................
```

> Some sycl header files depend on the c++17 feature, need to build with c++17.

Reviewer: ```suggestion

```

Timer                       Cumulative   Cumulative   Cumulative   Cumulative   Cumulative   Cumulative
Name                            number    microSecs    microSecs    microSecs    microSecs   Efficiency
                              of calls          min          avg          max       stddev       Rating
main                                 1    7.386e+07    7.386e+07    7.386e+07    0.000e+00       100.00
cycleInit                           10    2.754e+06    2.754e+06    2.754e+06    0.000e+00       100.00
cycleTracking                       10    7.110e+07    7.110e+07    7.110e+07    0.000e+00       100.00
cycleTracking_Kernel               107    7.093e+07    7.093e+07    7.093e+07    0.000e+00       100.00
cycleTracking_MPI                  117    1.751e+05    1.751e+05    1.751e+05    0.000e+00       100.00
cycleTracking_Test_Done              0    0.000e+00    0.000e+00    0.000e+00    0.000e+00         0.00
cycleFinalize                       20    3.690e+02    3.690e+02    3.690e+02    0.000e+00       100.00
Figure Of Merit                  12.67 [Num Mega Segments / Cycle Tracking Time]
```
**Note:** 

> * The testing result was running on Intel(R) Core(TM) i7-13700K CPU backend with Intel® oneAPI Base Toolkit(2023.2 version).

Reviewer: ```suggestion
* The testing result was collected when run on Intel(R) Core(TM) i7-13700K CPU backend with Intel® oneAPI Base Toolkit(2023.2 version).
```
